lib/string_map.h

```diff
@@ -0,0 +1,226 @@
+    using const_reverse_iterator = std::reverse_iterator<const_iterator>;
+
+ private:
+    using map_type = absl::flat_hash_map<cstring, list_iterator>;
+    map_type data_map;
+    void init_data_map() {
+        data_map.clear();
+        for (auto it = data.begin(); it != data.end(); it++) data_map.emplace(it->first, it);
+    }
+    iterator tr_iter(typename map_type::iterator i) {
+        if (i == data_map.end()) return data.end();
+        return i->second;
+    }
+    const_iterator tr_iter(typename map_type::const_iterator i) const {
+        if (i == data_map.end()) return data.end();
+        return i->second;
+    }
+
+ public:
+    using size_type = typename map_type::size_type;
+
+    string_map() = default;
+    string_map(const string_map &a) : data(a.data) { init_data_map(); }
+    template <typename InputIt>
+    string_map(InputIt first, InputIt last) {
+        insert(first, last);
+    }
+    string_map(string_map &&a) = default;
+    string_map &operator=(const string_map &a) {
+        /* std::list assignment broken by spec if elements are const... */
+        if (this != &a) {
+            data.clear();
+            data.insert(data.end(), a.data.begin(), a.data.end());
+            init_data_map();
+        }
+        return *this;
+    }
+    string_map &operator=(string_map &&a) = default;
+    string_map(std::initializer_list<value_type> il) { insert(il.begin(), il.end()); }
+
+    iterator begin() noexcept { return data.begin(); }
+    const_iterator begin() const noexcept { return data.begin(); }
+    iterator end() noexcept { return data.end(); }
+    const_iterator end() const noexcept { return data.end(); }
+    reverse_iterator rbegin() noexcept { return data.rbegin(); }
+    const_reverse_iterator rbegin() const noexcept { return data.rbegin(); }
+    reverse_iterator rend() noexcept { return data.rend(); }
+    const_reverse_iterator rend() const noexcept { return data.rend(); }
+    const_iterator cbegin() const noexcept { return data.cbegin(); }
+    const_iterator cend() const noexcept { return data.cend(); }
+    const_reverse_iterator crbegin() const noexcept { return data.crbegin(); }
+    const_reverse_iterator crend() const noexcept { return data.crend(); }
+
+    bool empty() const noexcept { return data.empty(); }
+    size_type size() const noexcept { return data_map.size(); }
+    size_type max_size() const noexcept { return data_map.max_size(); }
+    bool operator==(const string_map &a) const { return data == a.data; }
+    bool operator!=(const string_map &a) const { return data != a.data; }
+    void clear() {
+        data.clear();
+        data_map.clear();
+    }
+
+    // Functions above do have `std::string_view` versions. Here we are having
+    // important special case: if `a` is not something that was interned, we do
+    // not copy / intern it, we know for sure that `a` is not in the map and we
+    // do not need to perform a lookup
```

Comment:
The comment should probably be on the `string_view` version and it should be Doxygen.